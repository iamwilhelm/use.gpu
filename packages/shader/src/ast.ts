import { SyntaxNode } from '@lezer/common';

export const formatASTNode = (node: SyntaxNode) => {
  const {type} = node;
  let child = node.firstChild;
  let inner = [] as string[];
  while (child) {
    inner.push(formatASTNode(child));
    child = child.nextSibling;
  }
  const space = inner.length ? ' ' : '';
  return `(${type.name}${space}${inner.join(" ")})`;
}

export const formatAST = (node: SyntaxNode, program: string, depth: number = 0) => {
  const {type, from ,to} = node;
  const prefix = '  '.repeat(depth);

  let child = node.firstChild;
  
  const text = program.slice(node.from, node.to).replace(/\n/g, "‚Æê")
  let out = [] as string[];
  
  let line = `${prefix}${type.name}`;
  const n = line.length;
  line += ' '.repeat(60 - n);
  line += text;
  out.push(line);

  while (child) {
    out.push(formatAST(child, program, depth + 1));
    child = child.nextSibling;
  }
  return out.join("\n");
}
