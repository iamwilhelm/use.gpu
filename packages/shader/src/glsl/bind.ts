import { ShaderModule, DataBinding } from './types';

import { defineConstants } from './shader';
import { makeBindingAccessors, makeUniformBlock } from './gen';
import { makeResolveBindings, namespaceBinding } from '../util/bind';
import { VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP } from './constants';

export { bindBundle, bindModule } from '../util/bind';

const NO_SYMBOLS = [] as any[];

const getVirtualBindGroup = () => VIRTUAL_BINDGROUP;

export const bindingToModule = (
  binding: DataBinding,
): ShaderModule => {
  const {uniform: {name}} = binding;
  const links = makeBindingAccessors([binding], VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP);
  const module = links[name];
  return {...module, entry: name};
}

export const bindingsToLinks = (
  bindings: DataBinding[],
): Record<string, ShaderModule> => {
  return makeBindingAccessors(bindings, VIRTUAL_BINDGROUP, VOLATILE_BINDGROUP);
}

export const resolveBindings = makeResolveBindings(makeUniformBlock, getVirtualBindGroup);
