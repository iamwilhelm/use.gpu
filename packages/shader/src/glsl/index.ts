import {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

import {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,
} from './bind';

import {
  castTo,
  swizzleTo,
} from './cast';

import {
  chainTo,
} from './chain';

import {
  diffBy,
} from './diff';

import {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

import {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

import {
  getHash
} from '../util/hash';

import {
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

export {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,
} from './bind';

export {
  castTo,
  swizzleTo,
} from './cast';

export {
  chainTo,
} from './chain';

export {
  diffBy,
} from './diff';

export {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

export {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

export {
  getHash
} from '../util/hash';

export {
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export const GLSLLinker = {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  bindEntryPoint,
  bundleToAttribute,
  bundleToAttributes,
  glsl, float, int, uint,

  defineConstants,

  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,

  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  resolveBindings,

  castTo,
  chainTo,
  diffBy,
  swizzleTo,

  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,

  makeModuleCache,

  getHash,
  getBundleHash,
  getBundleKey,
};

export default GLSLLinker;