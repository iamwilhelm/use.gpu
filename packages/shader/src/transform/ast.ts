import { SyntaxNode, Tree } from '@lezer/common';
import {
  SymbolTable,
  SymbolRef,
  ModuleRef,
  ImportRef,
  PrototypeRef,
  FunctionRef,
  DeclarationRef,
  VariableRef,
  LocalRef,
  TypeRef,
  StructRef,
  MemberRef,
  QualifiedStructRef,
} from '../types';
import * as T from '../grammar/glsl.terms';
import { GLSL_NATIVE_TYPES, HASH_KEY } from '../constants';
import siphash from "../siphash13";

const unescape = (s: string) => s.slice(1, -1).replace(/\\(.)/g, '$1');

function nodeToString(this: SyntaxNode) { return `[${this.type.name}]`; }

export const hasErrorNode = (tree: Tree) => {
  const cursor = tree.cursor();
  do {
    const {type, from, to} = cursor;
    if (type.isError) return cursor.node;
  } while (cursor.next());
  return false;
}

export const getChildNodes = (node: SyntaxNode) => {
  const out = [] as SyntaxNode[];
  let n = node.firstChild;
  while (n) {
    // @ts-ignore
    n.__proto__.toJSON = nodeToString;
    out.push(n);
    n = n.nextSibling;
  }

  return out;
}

export const formatASTNode = (node: SyntaxNode) => {
  const {type} = node;
  let child = node.firstChild;
  let inner = [] as string[];
  while (child) {
    inner.push(formatASTNode(child));
    child = child.nextSibling;
  }
  const space = inner.length ? ' ' : '';
  return `(${type.name}${space}${inner.join(" ")})`;
}

export const formatAST = (node: SyntaxNode, program?: string, depth: number = 0) => {
  const {type, from ,to} = node;
  const prefix = '  '.repeat(depth);

  let child = node.firstChild;
  
  const text = program != null ? program.slice(node.from, node.to).replace(/\n/g, "⮐ ") : '';
  let out = [] as string[];
  
  let line = `${prefix}${type.name}`;
  const n = line.length;
  line += ' '.repeat(60 - n);
  line += text;
  out.push(line);

  while (child) {
    out.push(formatAST(child, program, depth + 1));
    child = child.nextSibling;
  }
  return out.join("\n");
}

export const makeASTParser = (code: string, tree: Tree) => {
  const throwError = (t: string, n?: SyntaxNode) => {
    if (!n) throw new Error(`Missing node`);
    console.log(formatAST(tree.topNode, code));
    throw new Error(`Error parsing ${t} node '${code.slice(n.from, n.to)}'\n${formatAST(n, code)}`);
  }

  const getNodes = (node: SyntaxNode, min?: number) => {
    const nodes = getChildNodes(node);
    for (const n of nodes) if (node.type.isError) throwError('error', node);
    if (min != null && nodes.length < min) throwError(`not enough nodes (${min})`, node);
    return nodes;
  }

  const getText = (node: SyntaxNode) => {
    if (!node) throwError('text');
    return code.slice(node.from, node.to);
  }
  
  const getQualifier = getText;
  
  const getQualifiedType = (node: SyntaxNode): TypeRef => {
    const [a, b] = getNodes(node);

    const qualifiers = b ? getNodes(a).map(getQualifier) : [];
    const def = b ?? a;

    const {name, members} = getTypeSpecifier(def);
    if (members) return {node, name, qualifiers, members};

    return {node, name, qualifiers};
  }
  
  const getTypeSpecifier = (node: SyntaxNode): TypeRef => {
    if (node.type.id === T.Identifier) {
      const name = getText(node);
      return {node, name};
    }
    if (node.type.id === T.TypeSpecifier) {
      const [def] = getNodes(node, 1);
      if (def.type.id === T.Struct) {
        const {name, members} = getStructType(def);
        return {node, name, members};
      }
      else {
        const name = getText(def);
        return {node, name};
      }
    }
    
    throw throwError('type specifier', node);
  }

  const getStructType = (node: SyntaxNode): TypeRef => {
    const [, b, c] = getNodes(node, 3);
    const name = getText(b);
    const {members} = getStructMembers(c);
    return {node, name, members};
  }

  const getArrayName = (node: SyntaxNode): string => {
    const [a] = getNodes(node, 1);
    return getText(a);
  }

  const getLocal = (node: SyntaxNode): LocalRef => {
    const [a, b] = getNodes(node, 1);
    const name = getArrayName(a);
    const expr = b ? getText(b) : null;
    return {node, name, expr};
  }

  const getMember = (node: SyntaxNode): MemberRef => {
    const [a, b] = getNodes(node, 2);
    const type = getTypeSpecifier(a);
    const name = getArrayName(b);
    return {node, name, type};
  }

  const getStructMembers = (node: SyntaxNode): StructRef => {
    const nodes = getNodes(node);
    const members = nodes.map(getMember);
    return {node, members};
  }

  const getImport = (node: SyntaxNode): ImportRef => {
    const [a, b] = getNodes(node, 1);
    const hasAlias = !!b;

    const imported = getText(a);
    const name = hasAlias ? getText(b) : imported;
    const symbol = { node, name };

    return {node, name, imported};
  };

  const getPrototype = (node: SyntaxNode): PrototypeRef => {
    const [a, b, c] = getNodes(node, 2);

    const type = getQualifiedType(a);
    const name = getText(b);
    const parameters = c ? getNodes(c).map(getText) : [];
    const symbols = [{ node, name }];

    return {node, name, type, parameters};
  };
  
  const getVariable = (node: SyntaxNode): VariableRef => {
    const [a, ...rest] = getNodes(node, 1);
    const type = getQualifiedType(a);
    const locals = rest.map(getLocal);

    return {node, type, locals};
  }
  
  const getQualifiedStruct = (node: SyntaxNode): QualifiedStructRef => {
    const [a, b, c, d] = getNodes(node, 3);
    
    const type = getQualifiedType(node);
    const name = getText(d ?? b);
    const struct = getStructMembers(c);

    return {node, name, type, struct};
  }
  
  const getFunction = (node: SyntaxNode): FunctionRef => {
    const [a] = getNodes(node, 1);
    const prototype = getPrototype(a);
    const symbols = [{node, name: prototype.name}];

    return {node, prototype, symbols};
  };

  const getDeclaration = (node: SyntaxNode): DeclarationRef => {
    const [a] = getNodes(node);

    if (a.type.id === T.FunctionPrototype) {
      const prototype = getPrototype(a);
      const {name} = prototype;
      const symbols = [{ node, name }];
      return {node, symbols, prototype};
    }
    if (a.type.id === T.VariableDeclaration) {
      const variable = getVariable(a);
      const symbols = variable.locals.map(({node, name}) => ({node, name}));

      const {type} = variable;
      if (!GLSL_NATIVE_TYPES.has(type.name)) symbols.push({node: type.node, name: type.name});

      return {node, symbols, variable};
    }
    if (a.type.id === T.QualifiedStructDeclaration) {
      const struct = getQualifiedStruct(a);
      const {name, type} = struct;

      const symbols = [{ node, name }];
      if (type.name === name) {
        // Struct is anonymous, members are global
        for (const {node, name, type} of struct.struct.members) {
          console.log('anon', name, type)
          symbols.push({node, name});
          if (!GLSL_NATIVE_TYPES.has(type.name)) symbols.push({node: type.node, name: type.name});
        }
      }
      else if (
        !GLSL_NATIVE_TYPES.has(type.name)
      ) {
        // Custom type
        symbols.push({node: type.node, name: type.name});
      }

      return {node, symbols, struct};
    }
    if (a.type.id === T.QualifiedDeclaration) {
      return {node, symbols: []};
    }
    
    throw throwError('declaration', node);
  };

  const isExported = (ref: FunctionRef | DeclarationRef): boolean => {  
    const {node} = ref;
    const pragma = node.prevSibling;

    if (!pragma || pragma.type.id !== T.Preprocessor) return false;

    const [a, b] = getNodes(pragma, 2);
    const directive = getText(a);
    if (directive !== 'pragma') return false;

    const verb = getText(b);
    return verb === 'export';
  };

  const extractImports = (): ModuleRef[] => {
    const modules: Record<string, ImportRef[]> = {};

    const children = tree.topNode.getChildren(T.Preprocessor);
    for (const child of children) {
      const [a, b, c, d] = getNodes(child);    
      const directive = getText(a);
      if (directive !== 'pragma') continue;

      const verb = getText(b);
      if (verb !== 'import') continue;

      const refs = getNodes(c).map(getImport);
      const module = unescape(getText(d));

      let items = modules[module];
      if (!items) items = modules[module] = [];
    
      items.push(...refs);
    }

    const out = [] as ModuleRef[];
    for (const k in modules) out.push({
      name: k,
      symbols: modules[k].map(({node, name}) => ({node, name})),
      imports: modules[k],
    });
    return out;
  }

  const extractFunctions = (): FunctionRef[] => {
    const children = tree.topNode.getChildren(T.FunctionDefinition);
    return children.map(getFunction);
  }

  const extractDeclarations = (): DeclarationRef[] => {
    const children = tree.topNode.getChildren(T.Declaration);
    return children.map(getDeclaration);
  };

  const extractSymbolTable = (): SymbolTable => {
    const uint = siphash.hash_uint(HASH_KEY, code);
    const hash = uint.toString(36).slice(-8);

    const modules = extractImports();
    const functions = extractFunctions();
    const declarations = extractDeclarations();

    const externals = declarations
      .filter(d => d.prototype)
      .filter(d => !functions.find(f => f.prototype.name === d.prototype!.name));

    const refs = [...modules, ...functions, ...declarations];
    const exported = [...functions, ...declarations].filter(isExported);
    
    const visibles = exported.flatMap(r => r.symbols);
    const symbols = refs.flatMap(r => r.symbols);

    return {hash, refs, symbols, visibles, externals, modules, functions, declarations};
  }
  
  return {extractImports, extractFunctions, extractDeclarations, extractSymbolTable};
}

export const rewriteAST = (code: string, tree: Tree, rename: Map<string, string>) => {
  let out = '';
  let pos = 0;

  const skip = (from: number, to: number, replace?: string) => {
    out = out + code.slice(pos, from);
    pos = to;

    if (replace != null) out = out + replace;
  }

  const cursor = tree.cursor();
  do {
    const {type, from, to} = cursor;
    if (type.name === 'version') {
      cursor.parent();
      cursor.parent();
      const {from, to} = cursor;
      skip(from, to, '');
      cursor.lastChild();
    }
    if (type.name === 'import' || type.name === 'export') {
      cursor.parent();
      const {from, to} = cursor;
      skip(from, to, '');
      cursor.lastChild();
    }
    else if (type.id === T.Identifier) {
      const name = code.slice(from, to);
      const replace = rename.get(name);

      if (replace) skip(from, to, replace);      
    }
  } while (cursor.next());

  const n = code.length;
  skip(n, n);

  return out;
}
