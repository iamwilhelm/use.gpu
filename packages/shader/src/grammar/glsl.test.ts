import { parser } from './glsl';
import { formatAST } from '../ast';

expect.addSnapshotSerializer({
  print(val) {
    return formatAST(val.topNode, val.text);
  },
  test(val) {
    return val && val.hasOwnProperty('type') && val.hasOwnProperty('children') && val.hasOwnProperty('positions');
  },
});

describe("GLSL grammar snapshots", () => {
  
  it("parses a test program", () => {
    for (const program of PROGRAMS) {
      const parsed = parser.parse(program);
      parsed.text = program;
      expect(parsed).toMatchSnapshot();
    }
  });
  
});

const PROGRAMS = [
`
float foo = 1.0;
#define WAT

struct light {
 float intensity;
 vec3 position;
} lightVar;
void main();
void main() {
  int bar = wat(5, 6);
  int x = 4 + 5 + +6;
  struct s { } x;
  gl_FragColor = vec4(0.1, 0.2, 0.3, 1.0);
}
`,
`
#version 450

layout(set = 0, binding = 0) uniform ViewUniforms {
  mat4 projectionMatrix;
  mat4 viewMatrix;
  vec4 viewPosition;
  vec4 lightPosition;
} view;

layout(location = 0) in vec4 position;
layout(location = 1) in vec4 normal;
layout(location = 2) in vec4 color;
layout(location = 3) in vec2 uv;

layout(location = 0) out vec4 fragColor;
layout(location = 1) out vec2 fragUV;

layout(location = 2) out vec3 fragNormal;
layout(location = 3) out vec3 fragLight;
layout(location = 4) out vec3 fragView;

void main() {
  gl_Position = view.projectionMatrix * view.viewMatrix * position;

  fragColor = color;
  fragUV = uv;

  fragNormal = normal.xyz;
  fragLight = view.lightPosition.xyz - position.xyz;
  fragView = view.viewPosition.xyz - position.xyz;
}
`
];

