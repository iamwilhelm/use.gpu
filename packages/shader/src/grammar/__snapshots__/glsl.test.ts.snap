// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GLSL grammar snapshots parses a test program 1`] = `
Program                                                     ⮐float foo = 1.0;⮐#define WAT⮐⮐struct light {⮐ float intensity;⮐ vec3 position;⮐} lightVar;⮐void main();⮐void main() {⮐  int bar = wat(5, 6);⮐  int x = 4 + 5 + +6;⮐  struct s { } x;⮐  gl_FragColor = vec4(0.1, 0.2, 0.3, 1.0);⮐}⮐
  Declaration                                               float foo = 1.0;
    VariableDeclaration                                     float foo = 1.0
      QualifiedType                                         float
        float                                               float
      SingleDeclaration                                     foo = 1.0
        DeclarationSpecifier                                foo
          Identifier                                        foo
        Assign                                              =
        Number                                              1.0
  Preprocessor                                              #define WAT
    Hash                                                    #
    Directive                                               define
      define                                                define
    UntilEOL                                                 WAT
  Declaration                                               struct light {⮐ float intensity;⮐ vec3 position;⮐} lightVar;
    VariableDeclaration                                     struct light {⮐ float intensity;⮐ vec3 position;⮐} lightVar
      QualifiedType                                         struct light {⮐ float intensity;⮐ vec3 position;⮐}
        Struct                                              struct light {⮐ float intensity;⮐ vec3 position;⮐}
          struct                                            struct
          Identifier                                        light
          StructDeclarationList                             float intensity;⮐ vec3 position;
            StructDeclaration                               float intensity;
              float                                         float
              DeclarationSpecifier                          intensity
                Identifier                                  intensity
            StructDeclaration                               vec3 position;
              vec3                                          vec3
              DeclarationSpecifier                          position
                Identifier                                  position
      SingleDeclaration                                     lightVar
        DeclarationSpecifier                                lightVar
          Identifier                                        lightVar
  Declaration                                               void main();
    FunctionPrototype                                       void main()
      QualifiedType                                         void
        void                                                void
      Identifier                                            main
  FunctionDefinition                                        void main() {⮐  int bar = wat(5, 6);⮐  int x = 4 + 5 + +6;⮐  struct s { } x;⮐  gl_FragColor = vec4(0.1, 0.2, 0.3, 1.0);⮐}
    FunctionPrototype                                       void main()
      QualifiedType                                         void
        void                                                void
      Identifier                                            main
    StatementList                                           int bar = wat(5, 6);⮐  int x = 4 + 5 + +6;⮐  struct s { } x;⮐  gl_FragColor = vec4(0.1, 0.2, 0.3, 1.0);
      Statement                                             int bar = wat(5, 6);
        DeclarationStatement                                int bar = wat(5, 6);
          Declaration                                       int bar = wat(5, 6);
            VariableDeclaration                             int bar = wat(5, 6)
              QualifiedType                                 int
                int                                         int
              SingleDeclaration                             bar = wat(5, 6)
                DeclarationSpecifier                        bar
                  Identifier                                bar
                Assign                                      =
                FunctionCall                                wat(5, 6)
                  Identifier                                wat
                  Number                                    5
                  Number                                    6
      Statement                                             int x = 4 + 5 + +6;
        DeclarationStatement                                int x = 4 + 5 + +6;
          Declaration                                       int x = 4 + 5 + +6;
            VariableDeclaration                             int x = 4 + 5 + +6
              QualifiedType                                 int
                int                                         int
              SingleDeclaration                             x = 4 + 5 + +6
                DeclarationSpecifier                        x
                  Identifier                                x
                Assign                                      =
                Binary                                      4 + 5 + +6
                  Binary                                    4 + 5
                    Number                                  4
                    Add                                     +
                    Number                                  5
                  Add                                       +
                  Unary                                     +6
                    Add                                     +
                    Number                                  6
      Statement                                             struct s { } x;
        DeclarationStatement                                struct s { } x;
          Declaration                                       struct s { } x;
            VariableDeclaration                             struct s { } x
              QualifiedType                                 struct s { }
                Struct                                      struct s { }
                  struct                                    struct
                  Identifier                                s
                  StructDeclarationList                     
              SingleDeclaration                             x
                DeclarationSpecifier                        x
                  Identifier                                x
      Statement                                             gl_FragColor = vec4(0.1, 0.2, 0.3, 1.0);
        ExpressionStatement                                 gl_FragColor = vec4(0.1, 0.2, 0.3, 1.0);
          Identifier                                        gl_FragColor
          Assign                                            =
          FunctionCall                                      vec4(0.1, 0.2, 0.3, 1.0)
            vec4                                            vec4
            Number                                          0.1
            Number                                          0.2
            Number                                          0.3
            Number                                          1.0
`;

exports[`GLSL grammar snapshots parses a test program 2`] = `
Program                                                     ⮐#version 450⮐⮐layout(set = 0, binding = 0) uniform ViewUniforms {⮐  mat4 projectionMatrix;⮐  mat4 viewMatrix;⮐  vec4 viewPosition;⮐  vec4 lightPosition;⮐} view;⮐⮐layout(location = 0) in vec4 position;⮐layout(location = 1) in vec4 normal;⮐layout(location = 2) in vec4 color;⮐layout(location = 3) in vec2 uv;⮐⮐layout(location = 0) out vec4 fragColor;⮐layout(location = 1) out vec2 fragUV;⮐⮐layout(location = 2) out vec3 fragNormal;⮐layout(location = 3) out vec3 fragLight;⮐layout(location = 4) out vec3 fragView;⮐⮐void main() {⮐  gl_Position = view.projectionMatrix * view.viewMatrix * position;⮐⮐  fragColor = color;⮐  fragUV = uv;⮐⮐  fragNormal = normal.xyz;⮐  fragLight = view.lightPosition.xyz - position.xyz;⮐  fragView = view.viewPosition.xyz - position.xyz;⮐}⮐
  Preprocessor                                              #version 450
    Hash                                                    #
    Directive                                               version
      version                                               version
    UntilEOL                                                 450
  Declaration                                               layout(set = 0, binding = 0) uniform ViewUniforms {⮐  mat4 projectionMatrix;⮐  mat4 viewMatrix;⮐  vec4 viewPosition;⮐  vec4 lightPosition;⮐} view;
    TypeQualifier                                           layout(set = 0, binding = 0) uniform
      LayoutQualifier                                       layout(set = 0, binding = 0)
        layout                                              layout
        LayoutQualifierId                                   set = 0
          Identifier                                        set
          Assign                                            =
          Number                                            0
        LayoutQualifierId                                   binding = 0
          Identifier                                        binding
          Assign                                            =
          Number                                            0
      StorageQualifier                                      uniform
        uniform                                             uniform
    Identifier                                              ViewUniforms
    StructDeclarationList                                   mat4 projectionMatrix;⮐  mat4 viewMatrix;⮐  vec4 viewPosition;⮐  vec4 lightPosition;
      StructDeclaration                                     mat4 projectionMatrix;
        mat4                                                mat4
        DeclarationSpecifier                                projectionMatrix
          Identifier                                        projectionMatrix
      StructDeclaration                                     mat4 viewMatrix;
        mat4                                                mat4
        DeclarationSpecifier                                viewMatrix
          Identifier                                        viewMatrix
      StructDeclaration                                     vec4 viewPosition;
        vec4                                                vec4
        DeclarationSpecifier                                viewPosition
          Identifier                                        viewPosition
      StructDeclaration                                     vec4 lightPosition;
        vec4                                                vec4
        DeclarationSpecifier                                lightPosition
          Identifier                                        lightPosition
    DeclarationSpecifier                                    view
      Identifier                                            view
  Declaration                                               layout(location = 0) in vec4 position;
    VariableDeclaration                                     layout(location = 0) in vec4 position
      QualifiedType                                         layout(location = 0) in vec4
        TypeQualifier                                       layout(location = 0) in
          LayoutQualifier                                   layout(location = 0)
            layout                                          layout
            LayoutQualifierId                               location = 0
              Identifier                                    location
              Assign                                        =
              Number                                        0
          StorageQualifier                                  in
            in                                              in
        vec4                                                vec4
      SingleDeclaration                                     position
        DeclarationSpecifier                                position
          Identifier                                        position
  Declaration                                               layout(location = 1) in vec4 normal;
    VariableDeclaration                                     layout(location = 1) in vec4 normal
      QualifiedType                                         layout(location = 1) in vec4
        TypeQualifier                                       layout(location = 1) in
          LayoutQualifier                                   layout(location = 1)
            layout                                          layout
            LayoutQualifierId                               location = 1
              Identifier                                    location
              Assign                                        =
              Number                                        1
          StorageQualifier                                  in
            in                                              in
        vec4                                                vec4
      SingleDeclaration                                     normal
        DeclarationSpecifier                                normal
          Identifier                                        normal
  Declaration                                               layout(location = 2) in vec4 color;
    VariableDeclaration                                     layout(location = 2) in vec4 color
      QualifiedType                                         layout(location = 2) in vec4
        TypeQualifier                                       layout(location = 2) in
          LayoutQualifier                                   layout(location = 2)
            layout                                          layout
            LayoutQualifierId                               location = 2
              Identifier                                    location
              Assign                                        =
              Number                                        2
          StorageQualifier                                  in
            in                                              in
        vec4                                                vec4
      SingleDeclaration                                     color
        DeclarationSpecifier                                color
          Identifier                                        color
  Declaration                                               layout(location = 3) in vec2 uv;
    VariableDeclaration                                     layout(location = 3) in vec2 uv
      QualifiedType                                         layout(location = 3) in vec2
        TypeQualifier                                       layout(location = 3) in
          LayoutQualifier                                   layout(location = 3)
            layout                                          layout
            LayoutQualifierId                               location = 3
              Identifier                                    location
              Assign                                        =
              Number                                        3
          StorageQualifier                                  in
            in                                              in
        vec2                                                vec2
      SingleDeclaration                                     uv
        DeclarationSpecifier                                uv
          Identifier                                        uv
  Declaration                                               layout(location = 0) out vec4 fragColor;
    VariableDeclaration                                     layout(location = 0) out vec4 fragColor
      QualifiedType                                         layout(location = 0) out vec4
        TypeQualifier                                       layout(location = 0) out
          LayoutQualifier                                   layout(location = 0)
            layout                                          layout
            LayoutQualifierId                               location = 0
              Identifier                                    location
              Assign                                        =
              Number                                        0
          StorageQualifier                                  out
            out                                             out
        vec4                                                vec4
      SingleDeclaration                                     fragColor
        DeclarationSpecifier                                fragColor
          Identifier                                        fragColor
  Declaration                                               layout(location = 1) out vec2 fragUV;
    VariableDeclaration                                     layout(location = 1) out vec2 fragUV
      QualifiedType                                         layout(location = 1) out vec2
        TypeQualifier                                       layout(location = 1) out
          LayoutQualifier                                   layout(location = 1)
            layout                                          layout
            LayoutQualifierId                               location = 1
              Identifier                                    location
              Assign                                        =
              Number                                        1
          StorageQualifier                                  out
            out                                             out
        vec2                                                vec2
      SingleDeclaration                                     fragUV
        DeclarationSpecifier                                fragUV
          Identifier                                        fragUV
  Declaration                                               layout(location = 2) out vec3 fragNormal;
    VariableDeclaration                                     layout(location = 2) out vec3 fragNormal
      QualifiedType                                         layout(location = 2) out vec3
        TypeQualifier                                       layout(location = 2) out
          LayoutQualifier                                   layout(location = 2)
            layout                                          layout
            LayoutQualifierId                               location = 2
              Identifier                                    location
              Assign                                        =
              Number                                        2
          StorageQualifier                                  out
            out                                             out
        vec3                                                vec3
      SingleDeclaration                                     fragNormal
        DeclarationSpecifier                                fragNormal
          Identifier                                        fragNormal
  Declaration                                               layout(location = 3) out vec3 fragLight;
    VariableDeclaration                                     layout(location = 3) out vec3 fragLight
      QualifiedType                                         layout(location = 3) out vec3
        TypeQualifier                                       layout(location = 3) out
          LayoutQualifier                                   layout(location = 3)
            layout                                          layout
            LayoutQualifierId                               location = 3
              Identifier                                    location
              Assign                                        =
              Number                                        3
          StorageQualifier                                  out
            out                                             out
        vec3                                                vec3
      SingleDeclaration                                     fragLight
        DeclarationSpecifier                                fragLight
          Identifier                                        fragLight
  Declaration                                               layout(location = 4) out vec3 fragView;
    VariableDeclaration                                     layout(location = 4) out vec3 fragView
      QualifiedType                                         layout(location = 4) out vec3
        TypeQualifier                                       layout(location = 4) out
          LayoutQualifier                                   layout(location = 4)
            layout                                          layout
            LayoutQualifierId                               location = 4
              Identifier                                    location
              Assign                                        =
              Number                                        4
          StorageQualifier                                  out
            out                                             out
        vec3                                                vec3
      SingleDeclaration                                     fragView
        DeclarationSpecifier                                fragView
          Identifier                                        fragView
  FunctionDefinition                                        void main() {⮐  gl_Position = view.projectionMatrix * view.viewMatrix * position;⮐⮐  fragColor = color;⮐  fragUV = uv;⮐⮐  fragNormal = normal.xyz;⮐  fragLight = view.lightPosition.xyz - position.xyz;⮐  fragView = view.viewPosition.xyz - position.xyz;⮐}
    FunctionPrototype                                       void main()
      QualifiedType                                         void
        void                                                void
      Identifier                                            main
    StatementList                                           gl_Position = view.projectionMatrix * view.viewMatrix * position;⮐⮐  fragColor = color;⮐  fragUV = uv;⮐⮐  fragNormal = normal.xyz;⮐  fragLight = view.lightPosition.xyz - position.xyz;⮐  fragView = view.viewPosition.xyz - position.xyz;
      Statement                                             gl_Position = view.projectionMatrix * view.viewMatrix * position;
        ExpressionStatement                                 gl_Position = view.projectionMatrix * view.viewMatrix * position;
          Identifier                                        gl_Position
          Assign                                            =
          Binary                                            view.projectionMatrix * view.viewMatrix * position
            Binary                                          view.projectionMatrix * view.viewMatrix
              Field                                         view.projectionMatrix
                Identifier                                  view
                Identifier                                  projectionMatrix
              Mul                                           *
              Field                                         view.viewMatrix
                Identifier                                  view
                Identifier                                  viewMatrix
            Mul                                             *
            Identifier                                      position
      Statement                                             fragColor = color;
        ExpressionStatement                                 fragColor = color;
          Identifier                                        fragColor
          Assign                                            =
          Identifier                                        color
      Statement                                             fragUV = uv;
        ExpressionStatement                                 fragUV = uv;
          Identifier                                        fragUV
          Assign                                            =
          Identifier                                        uv
      Statement                                             fragNormal = normal.xyz;
        ExpressionStatement                                 fragNormal = normal.xyz;
          Identifier                                        fragNormal
          Assign                                            =
          Field                                             normal.xyz
            Identifier                                      normal
            Identifier                                      xyz
      Statement                                             fragLight = view.lightPosition.xyz - position.xyz;
        ExpressionStatement                                 fragLight = view.lightPosition.xyz - position.xyz;
          Identifier                                        fragLight
          Assign                                            =
          Binary                                            view.lightPosition.xyz - position.xyz
            Field                                           view.lightPosition.xyz
              Field                                         view.lightPosition
                Identifier                                  view
                Identifier                                  lightPosition
              Identifier                                    xyz
            Sub                                             -
            Field                                           position.xyz
              Identifier                                    position
              Identifier                                    xyz
      Statement                                             fragView = view.viewPosition.xyz - position.xyz;
        ExpressionStatement                                 fragView = view.viewPosition.xyz - position.xyz;
          Identifier                                        fragView
          Assign                                            =
          Binary                                            view.viewPosition.xyz - position.xyz
            Field                                           view.viewPosition.xyz
              Field                                         view.viewPosition
                Identifier                                  view
                Identifier                                  viewPosition
              Identifier                                    xyz
            Sub                                             -
            Field                                           position.xyz
              Identifier                                    position
              Identifier                                    xyz
`;
